### ISSUE TITLE
Define a canonical, reproducible order of execution

### ISSUE THREAD

**Issue by jonrkarr** [SED-ML/sed-ml#115]
To support reproducible execution of Monte Carlo algorithms, there should be a defined sorting algorithm which establishes a reproducible order of execution irrespective of the order in which element are defined in XML files. 

One option is to simply use the order in the XML file, but this runs counter to the XML order having no semantic meaning. 

A preferable option is to sort using the values of ids and other attributes.

**Comment by matthiaskoenig** [comment]
I strongly disagree. Similar to #117 this is not an issue for reproducibility. 
Enforcing a canonical order is a killer for parallelization and scaling on multiple cores/machines. The content of the data generators must be correct, how tools achieve this and in which order is up to the implementation. The content of deterministic data generators must be reproducible, the content of stochastic (or anything depending on randomness) is not-reproducible by result. But every tool will give different numerical results. A single step/run of a Monte Carlo algorithm is by definition not-reproducible, only the moments of ensemble runs are reproducible (a good source of information are the stochastic test cases for SBML https://github.com/sbmlteam/sbml-test-suite/blob/master/cases/stochastic/README.md with the respective PDF mentioned in the README).

The only thing which must be reproducible are the moments of the distributions for repeated stochastic simulations.
In summary: a canonical order will not solve the issue of reproducible random simulations, but inflict strong restrictions which hamper implementations.

**Comment by jonrkarr** [comment]
It is possible to have replicable executions even with parallel execution. But this does take work. The point of this issue is not to force tools to numerically replicable with each other. 

The point is to promote a shared understanding of what replicability means, mainly within each tool. This would for example, enable automated check of whether each tool is replicable within itself. From a strict computer science perspective, SED-ML presently stops short of prescribing replicability. This leaves it up to each implementation of SED-ML to define what replicability means.

**Comment by luciansmith** [comment]
I also don't think we need to define in SED-ML what tools need to do to accomplish their own repeatability.  If they do the same thing they did the last time, whatever that is, they will achieve repeatability.  

It is probably worth putting something in the spec that mentions this explicitly; something like "When a stochastic run is given a seed, simulators should be aware that the user expects to get the same results from repeated runs on the same architecture, even if some tasks are run in parallel."

**Comment by luciansmith** [comment]
Done!  See https://github.com/SED-ML/sed-ml/issues/117 for more discussion on the wording.
